### PR TITLE
Reimplementation of InDesign handler

### PR DESCRIPTION
**Pull request by kbairak**
Problem and/or solution
-----------------------

It turns out that using UCFLib can damage the idml archive. The
indications we have of this is that files compiled by this handler:

1. Show up as "corrupted" in InDesign
2. Have messed up permissions when extracted by archive managers (tested
   on Linux and MacOS operating systems)

The new implementation operates on the archives directly. We access
'designmap.xml' to gather the story IDs, then gather all the story files
and extract the content their <Content> tags as templates. Finally we
create a new archive identical in every way possible to the old one but
with the stories replaced by the templates. A very similar process is
performed during compilation.

It turns out that trying to retain as much of the structure and content
of the resulting archive fixes our issues.

How to test
-----------

Theoretically, the output of parsing and compiling, at least in terms of
strings and individual story templates, should be the same as before,
thus the same tests should pass.

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following [best practices](https://tem.transifex.com)
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message [according to TEM](https://tem.transifex.com/github-guide.html#working-on-a-feature)

### CODE REVIEW

File: openformats/formats/indesign.py [hunk 1]
```diff
@@ -122,103 +121,92 @@ def _can_skip_content(self, string):
...
+                transcriber.skip(len(self.string.template_replacement))
+                self.string = next(self.stringset, None)
+
+        transcriber.copy_to_end()
+        compiled = transcriber.get_destination()
+        compiled = hash_regex.sub(u'', compiled)
+        return compiled
+
+    @staticmethod
+    def _get_story_filenames(z_in):
```
Comment (rigaspapas) [inline openformats/formats/indesign.py]: Same here. I prefer to rename `z_in` to something like `zip_file`

File: openformats/formats/indesign.py [hunk 2]
```diff
@@ -122,103 +121,92 @@ def _can_skip_content(self, string):
...
+        """ Find the filenames of all Story fragmens by intersecting the list
+            of keys of the 'StoryList' attribute of the top node in
+            'designmap.xml' with the filenames within the archive that refer to
+            Story fragments.
```
Comment (rigaspapas) [inline openformats/formats/indesign.py]: It is not clear for me what is the goal of this method based on the docstring. Is it to remove the stories that are not referenced in `designmap.xml`? Is it the opposite? Can you explicitly mention it?

File: openformats/formats/indesign.py [hunk 3]
```diff
@@ -122,103 +121,92 @@ def _can_skip_content(self, string):
...
+              much of the structure and content as possible, but replacing the
+              story fragments with the compiled ones
+            * Return the compiled archive
+        """
 
-        out = io.BytesIO()
-        idml.save(out)
-        return out.getvalue()
+        self.stringset, compiled_dict = iter(stringset), {}
+        self.string = next(self.stringset)
```
Comment (rigaspapas) [inline openformats/formats/indesign.py]: So, in order for _compile_story to work, we need to initialize the `self.string` to the first item of the `self.stringset` list. Is this correct? Why not make `_compile_story` iterate through `self.stringset` and depend on an external initialization?

File: openformats/formats/indesign.py (hunk 2 above)
Comment (kbairak) [reply openformats/formats/indesign.py]: I will rewrite this hoping it's better next time

File: openformats/formats/indesign.py (hunk 3 above)
Comment (kbairak) [reply openformats/formats/indesign.py]: This comes up fairly regularly in openformats. The reason is that you are trying to iterate over a stringset and the "sections" of a template at the same time. Also, some sections do not correspond to strings because they may have been removed from the stringset because the string is untranslated. After having written a few handlers, I have concluded that `stringset = iter(stringset); string = next(stringset, None)` is the most convenient approach.

If you look at the old implementation, it did something like:

```python
for section in sections:
    string = stringset.pop(0)
    if string_matches_section(string, section):
        add_string_to_section(string, section)
    else:
        stringset.insert(string, 0)
```

I think that having the next string at the ready, adding checks if it's `None` (because we ran out of strings before we finished all the sections) and fetching the next string if we use the current string when compiling is the simplest to read and fastest approach.
